src/main/java/org/opensearch/neuralsearch/query/HybridBulkScorer.java

```diff
@@ -0,0 +1,189 @@
+    private static final int SHIFT = 12;
+    private static final int WINDOW_SIZE = 1 << SHIFT;
+    private static final int MASK = WINDOW_SIZE - 1;
+
+    private final long cost;
+    private final Scorer[] scorers;
+    @Getter
+    private final HybridSubQueryScorer hybridSubQueryScorer;
+    private final boolean needsScores;
+    @Getter
+    private final FixedBitSet matching;
+    @Getter
+    private final float[][] windowScores;
+    private final HybridQueryDocIdStream hybridQueryDocIdStream;
+    private final int maxDoc;
+
+    /**
+     * Constructor for HybridBulkScorer
+     * @param scorers list of scorers for each sub query
+     * @param needsScores whether the scorer needs scores
+     * @param maxDoc maximum document id
+     */
+    public HybridBulkScorer(List<Scorer> scorers, boolean needsScores, int maxDoc) {
+        long cost = 0;
+        this.scorers = new Scorer[scorers.size()];
+        for (int subQueryIndex = 0; subQueryIndex < scorers.size(); subQueryIndex++) {
+            Scorer scorer = scorers.get(subQueryIndex);
+            if (Objects.isNull(scorer)) {
+                continue;
+            }
+            cost += scorer.iterator().cost();
+            this.scorers[subQueryIndex] = scorer;
+        }
+        this.cost = cost;
+        this.hybridSubQueryScorer = new HybridSubQueryScorer(scorers.size());
+        this.needsScores = needsScores;
+        this.matching = new FixedBitSet(WINDOW_SIZE);
+        this.windowScores = new float[this.scorers.length][WINDOW_SIZE];
+        this.maxDoc = maxDoc;
+        this.hybridQueryDocIdStream = new HybridQueryDocIdStream(this);
+    }
+
+    @Override
+    public int score(LeafCollector collector, Bits acceptDocs, int min, int max) throws IOException {
+        collector.setScorer(hybridSubQueryScorer);
+        // making sure we are not going over the global limit defined by maxDoc
+        max = Math.min(max, maxDoc);
+        int[] docsIds = advance(min, scorers);
+        while (allDocIdsUsed(docsIds, max) == false) {
+            scoreWindow(collector, acceptDocs, min, max, docsIds);
+        }
+        return getNextDocIdCandidate(docsIds);
+    }
+
+    private void scoreWindow(LeafCollector collector, Bits acceptDocs, int min, int max, int[] docIds) throws IOException {
+        // pick the lowest out of all not yet used doc ids
+        int topDoc = -1;
+        for (int docId : docIds) {
+            if (docId < max) {
+                topDoc = docId;
+                break;
+            }
+        }
+
+        final int windowBase = topDoc & ~MASK; // take the next match (at random) and find the window where it belongs
+        final int windowMin = Math.max(min, windowBase);
+        final int windowMax = Math.min(max, windowBase + WINDOW_SIZE);
+
+        scoreWindowIntoBitSetWithSubqueryScorers(collector, acceptDocs, max, docIds, windowMin, windowMax, windowBase);
+    }
+
+    private void scoreWindowIntoBitSetWithSubqueryScorers(
```

Comment:
Please explain at high level what is happening in following method in the form of comment.

Reply:
Is this method similar to scoreAllRange method of bulkScorer?

Reply:
> Is this method similar to scoreAllRange method of bulkScorer?

not really, it collects ids and scores for all sub queries using leaf collector, but only within the boundary of the window of size 2^12

Reply:
ack adding more comments